DM-7433: "Create test case for processFile"

wmwv opened the pull request:
Add just a basic test that processFile.py runs and produces non-empty calexp, catalog, and calibcatalog files.

wmwv committed:
Start tests directory with SConscript.

wmwv committed:
Add testProcessFile to test basic processing.

processFile.py is run in the setUp method.
and then individual tests in the class can test for non-empty output files
for the outputCalex, outputCatalog, and outputCalibCatalog
and, eventually, more sophisticated tests.

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
+
```
Per the recent RFC-215, please remove this shabang and make the file non-executable.

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+    testDataDirectory = None
+
+
+class DoesProcessFileRun(unittest.TestCase):
```
Generally TestCases have "Test" in their name somewhere". So, `TestProcessFileRun` or `ProcessFileRunTest`

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+
+class DoesProcessFileRun(unittest.TestCase):
+    """Test that processFile runs."""
+    # Ideally processFile.py would just be a call to
```
Long comments like this should probably just be part of the docstring: there's useful information in the comment, and that information may help someone debugging a problem.

parejkoj replied:
Ah! Now that I've looked at runSampleProcessFile, I think you should move this comment to a slightly-reworded docstring on top of that method. It's actually not a bad thing to have the test do the running explicitly via subprocess: that's how users are going to use it, so I'm not sure the "ideally" bit is necessary.

parejkoj commented:
You'll want to make afwdata an optional dependency in the eups table `SetupOptional(afwdata)`, otherwise the new test will never be run automatically.

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+
+        # We run processFile.py here in the setUp method.
+        # so that the results are availalbe to the individual tests
+        self.runSampleProcessFile()
```
You shouldn't generally run anything during setUp: configure all the filenames and such, and then do the "run" step in a test. If that means you only have one test method `testRunProcessFile()`with three asserts in it based on what should have happened, that's ok (though not ideal: the "one test-one assert" design is very nice, doesn't always quite fit).

wmwv replied:
It takes about 45 seconds for the image to be processed, so keeping them all in one test is the only way to go if we don't put the running in `setUp`.

While the current existing tests are trivial and would even logically make sense in one test, I structured things this way with the set up so that adding future tests (as you suggest) of looking more deeply at the actual contents of the test would be civilized.

We are on course to have 6-12 tests here.  Should they all be in the same test function?

[This discussion is exactly why I asked you to review this ticket.]

jdswinbank replied:
Isn't `setUp()` called once per test _method_, not once per test case? That is, putting this in `setUp()` would still run it for _each_ of your 6-12 tests. If you want something that runs once per test case, so that you can use the result in all of the tests without recomputing it, you need `setUpClass()`.

(And personally, I don't see anything wrong with running this in `setUpClass()` given the circumstances, but I'll let you argue that one with Other John.)

wmwv replied:
Yes!  Thank you.  `setUpClass` was what was in my brain but not in my code.

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+        self.runSampleProcessFile()
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpPath)
```
I'd wrap this in an `if os.exists(self.tmpPath)`, otherwise you can get strange exceptions if something fails to create that directory.

parejkoj commented on tests/testProcessFile.py:
```diff
@@ -0,0 +1,105 @@
+        return output
+
+    def assertFileNotEmpty(self, pathname):
+        sizeOfFile = os.stat(pathname).st_size
```
Do you know something about what these files should actually contain? Even checking a handful of values or headers in them would be a step up.

parejkoj commented:
Once you've moved `runSampleProcessFile()` from `setUp()` into a separate test, you can also assert that the process returned 0 and/or that stdout/stderr contain whatever is appropriate.

parejkoj commented:
I'd also suggest rebasing to flatten several of these commits together.

parejkoj commented:
If you want each `test` method to have fewer asserts in it, you could lift the `--object*` stuff into their own test, and have `runSampleProcessFile` take arguments that get passed to subprocess for future things that you want to test. You'd then have `testOutputCalib` which runs with `--outputCalib blah`, `testOutputCalexp` which runs with `--outputCalexp blah`, etc. I don't know if this is feasible based on how the different cmdline arguments interact.

wmwv committed:
Add afwdata as setupOptional to support testing.

wmwv committed:
Alphabetize import lsst

wmwv committed:
Test that calexp image and catalogs are of reasonable size.

The thresholds are class attributes
self.expectedNumberOfObjects
self.expectedNumberOfCalibObjects
self.expectedSizeOfImage

These were previously defaults for keyword arguments.
It makes more sense to instead keep them with the class setUp
so that
(a) they're next to the definitions of the file to process
(b) it's more clear how to override them and
    potentially re-use tests across a range of images.

wmwv commented:
Thank you, @parejkoj.  I've improved the testing to check for output calexp size and number of entries in the catalogs, as well as addressed the absent `setupOptional(afwdata)` (doh!), squashing some commits, and other minor points.

The major philosophical distinction remains.  There is still not a clear test of "this runs".  Instead, a failure to run will be revealed by one of the individual tests reporting failure.  But for now we lack a clear integration test framework and having this at least covered will be much better than not covered at all.  Conversation on this topic continuing at 
https://community.lsst.org/t/should-image-process-tests-be-run-in-setupclass-so-results-are-available-to-all-tests/1105/11
